storage-bigtable/src/access_token.rs

```diff
@@ -91,24 +92,35 @@ impl AccessToken {
 
     /// Call this function regularly to ensure the access token does not expire
     pub async fn refresh(&self) {
-        // Check if it's time to try a token refresh
+        //Check if it's time to try a token refresh
         {
             let token_r = self.token.read().unwrap();
+
             if token_r.1.elapsed().as_secs() < token_r.0.expires_in() as u64 / 2 {
+                info!("Token not ready to be refreshed");
                 return;
             }
+            warn!("Token ready to be refreshed");
```

Comment:
nit: warn is far too chatty for these new log messages. if they need to stay, please demote to debug

Reply:
> can you elaborate on the root cause and how this change fixes it?
> 
> seems like at best it allows multiple outstanding refresh requests, which could stomp on each other

The root cause appears to be that the request to get a new token does not fail and continues effectively forever (or is garbage collected) and does not revert the refresh_active bool flag to false as a result, this may indeed be a memory leak (processes continuing to build up with no end) but I'm not a Rust dev so I can't tell, certainly if the GoAuth was failing, it should be caught, and there is no error message in the logs.

The current code, in the [second test](https://github.com/solana-labs/solana/blob/b5e0e83ed2cc54bbfb197ce991706c509e65bf0f/storage-bigtable/src/access_token.rs#L104) (after testing elapsed time for the token), falls out to renew the token and at the same time switches refresh_active to true (.compare_and_swap(false, true, Ordering::Relaxed)), but because the call does not complete, it never gets set back to false as it should [later in the call ](https://github.com/solana-labs/solana/blob/b5e0e83ed2cc54bbfb197ce991706c509e65bf0f/storage-bigtable/src/access_token.rs#L119)

I first commented out both tests and determined that even requesting a new token with every single call on a live production RPC was returning very quickly, but obviously this would not be ideal, but gave me the information that a successful auth request is very quick (less than 2 seconds but I am being conservative here).

My fix changes the behaviour and I think is quite sane, it doesn't affect any other successful calls because if the token is changed by a parallel call the token is updated anyway and the value refresh_active is changed to false anyway (second link) all my change does is says, if refresh_active is true, wait for 2 seconds (by which time it should complete) and change the value to false for the next call and return (it does not attempt to renew the token). This = that if the renew call has failed, it will not block the next renew call.

As I wrote in the comments, I am sure there are better approaches, but this issue has been there since September at least and nobody had fixed it, my method does indeed fix it. One better approach would be to actually make the renew request time out by itself, if there is no response, success or failure, within 2 seconds it should just gracefully exit and set refresh_active to false, but that is a lot more code and as I said, I am not a Rust dev.

Feel free to make a different fix for it, I have applied this fix to 25 active production RPC nodes for Ankr and it is working just fine.

Reply:
> The root cause appears to be that the request to get a new token sometimes does not succeed or fail and continues effectively forever (or is garbage collected) and does not revert the refresh_active bool flag to false as a result

This is not a correct statement.
Even if the request to get a new token fails flag still will be set to `false` because we do not `unwrap/expect` result, we handle `Ok/Err` with `match`.
I also do not think that process continues effectively forever, and as you tested new token returned very quickly each time.

The code looks completely ok, what is only possible are multiple updates but this should be ok and can be solved with `Ordering::SeqCst` memory order.

--
I do not understand why getting a new token is a stuck :neutral_face:

Reply:
> > The root cause appears to be that the request to get a new token sometimes does not succeed or fail and continues effectively forever (or is garbage collected) and does not revert the refresh_active bool flag to false as a result
> 
> This is not a correct statement. Even if the request to get a new token fails flag still will be set to `false` because we do not `unwrap/expect` result, we handle `Ok/Err` with `match`. I also do not think that process continues effectively forever, and as you tested new token returned very quickly each time.
> 
> The code looks completely ok, what is only possible are multiple updates but this should be ok and can be solved with `Ordering::SeqCst` memory order.
> 
> -- I do not understand why getting a new token is a stuck 😐

You are declaring this without testing though right? because in checking the logging, the code is not getting past the second test. That indicates that the value is true, continuously.

Reply:
Also in testing the token DOES NOT get returned every time quickly, it can fail continuously in testing up to 34 seconds in row.

Reply:
Yes, I'm declaring this without testing.

> Also in testing the token DOES NOT get returned every time quickly, it can fail continuously in testing up to 34 seconds in row.

Again, even if getting a new token fails, the flag still will be switched to `false`.

Reply:
"Yes, I'm declaring this without testing."

Reply:
"Again, even if getting a new token fails, the flag still will be switched to false."

I did not say that the token refresh returns an error, failing in this context means "it did not refresh the token". In my testing, there is no error, there is no logging event that indicates that any value, success or failure, is returned from the GoogleAuth request, and for an extended period of time multiple attempts to refresh happen, which are logged (if my full patch is applied, and the value reset to false in the second test block) and still, there is no error event logged during this time by the existing code.

If the solution is to change the ordering, great, would love it to be a simpler solution and not have to use my patch.

Reply:
Am I understanding correctly that in logs you see:
```
        info!("Requesting token for {:?} scope", scope);
```
but do not see
```
        info!("Token expires in {} seconds", token.expires_in());
```
?

Reply:
@PeaStew I patched solana sources a little (make internal things public) and started this code, everything looks good for now, tokens keep creating. We will try to determine the issue from the validator logs but not sure about this honestly :neutral_face: 
```rust
use {
    solana_storage_bigtable::{
        access_token::{AccessToken, Scope},
        CredentialType,
    },
    std::sync::Arc,
    tokio::time,
    tracing_subscriber::prelude::*,
};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    tracing_subscriber::registry()
        .with(tracing_subscriber::fmt::layer())
        .with(tracing_subscriber::filter::LevelFilter::INFO)
        .init();
    log::info!("logs will be here");

    let at = AccessToken::new(Scope::BigTableDataReadOnly, CredentialType::Filepath(None))
        .await
        .unwrap();
    let at = Arc::new(Box::new(at));

    loop {
        {
            let mut at = at.token.write().unwrap();
            (*at).1 -= time::Duration::from_secs(3600);
        }

        let at1 = Arc::clone(&at);
        let at2 = Arc::clone(&at);

        let _ = tokio::join!(
            tokio::spawn(async move { at1.refresh().await }),
            tokio::spawn(async move { at2.refresh().await }),
        );

        log::info!("---");
        time::sleep(time::Duration::from_secs(1)).await;
    }

    time::sleep(time::Duration::from_secs(3600)).await;

    Ok(())
}
```

Reply:
Hey @PeaStew, did you have a chance to test my patch?

Reply:
@fanatid , thanks for the patch, looks pretty reasonable. I'm thinking to ask one of the foundation rpc nodes to try it out, since we haven't heard anything from PeaStew

Reply:
I'm updated code a little, so no `unwrap` now: https://github.com/fanatid/solana/commit/32afd20b40a2cf4d3e3e2ba9fdc0f326818d841f

Reply:
Is this patch still good to try? We experience this same issue multiple times per day, so we could probably report back.